planner/cascades/transformation_rules.go

```diff
@@ -2315,3 +2316,66 @@ func (r *PullSelectionUpApply) OnTransform(old *memo.ExprIter) (newExprs []*memo
 	newApplyGroupExpr.SetChildren(outerChildGroup, old.Children[1].GetExpr().Children[0])
 	return []*memo.GroupExpr{newApplyGroupExpr}, false, false, nil
 }
+
+// PushProjDownUnionAll pushes the Projection down to the child of UnionAll.
+type PushProjDownUnionAll struct {
+	baseRule
+}
+
+// NewRulePushProjDownUnionAll creates a new Transformation PushProjDownUnionAll.
+// The pattern of this rule is: `Proj -> UnionAll)`.
+func NewRulePushProjDownUnionAll() Transformation {
+	rule := &PushProjDownUnionAll{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandProjection,
+		memo.EngineTiDBOnly,
+		memo.NewPattern(memo.OperandUnionAll, memo.EngineTiDBOnly),
+	)
+	return rule
+}
+
+// Match implements Transformation interface.
+func (r *PushProjDownUnionAll) Match(expr *memo.ExprIter) bool {
+	if len(expr.Children[0].GetExpr().Children) <= 1 {
+		return false
+	}
+
+	proj := expr.GetExpr().ExprNode.(*plannercore.LogicalProjection)
+	for _, expr := range proj.Exprs {
```

Comment:
Why does `proj.Exprs` need to contain ScalarFunction? IMHO, push Projection through UnionAll is always better. Do you have other concerns?

Reply:
push column is useless, and push constant is wrose.
```
select a, b from (
	select a, 1 as b from t
	union all
	select a, 1 as b from t
) t
```
is wrose than
```
select a, 1 from (
	select a from t
	union all
	select a from t
) t
```

Reply:
So I should just push down column and scalaFunc, and retain constant in `topProj`.
if there are no constant, `topProj` can be eliminated?

Reply:
I think the simplification of this rule is much more important than this performance issue. 
We can accept the performance regression of pushing down the constant here(and it won't be much).

Reply:
Assume that the cost of reading a col from chunk is 1 and writing a col to chunk is 1.

We can calculate the cost of projection<`k*col, h*constant`> for `n` rows is `(k + (k+h)) * n = (2k+h)*n`.
And the cost of union_all<`k*col`> for `n` rows is `2k*n`.

So the cost of 
```
projection<k*col, h*constant>
			|
	union_all<k*col>
			|
	projection<k*col>
```
is `(2k+h)*n + 2k*n + 2k*n = (6k+h)*n`.

And after we push down the topproj,
the cost of 
```
	union_all<(k+h)*col>
			|
	projection<k*col, h*constant>
```
is `2*(k+h)*n + (2k+h)*n = (4k+3h)*n`

`cost1 - cost2 = (6k+h)*n - (4k+3h)*n = 2n*(k-h)`.

when `k<h`, push down topproj has a worse cost.

if there are to many rows, the affect may be great?
And times(p) of push_down_proj applied also affect the cost?
The final gap is `2n*(k-h)*p`.

Reply:
if hasScalarFunction || (len(constant) <= len(column)) {
      push_down.
}
like this?